tests/nsga2.cpp

```diff
@@ -199,3 +199,34 @@ BOOST_AUTO_TEST_CASE(nsga2_serialization_test)
         }
     }
 }
+
+BOOST_AUTO_TEST_CASE(bfe_usage_test)
+{
+    // 1 - Algorithm with bfe disabled
+    problem prob{dtlz(1, 10, 2)};
+    nsga2 uda1{nsga2{100}};
+    uda1.set_verbosity(1u);
+    uda1.set_seed(23u);
+    // 2 - Instantiate
+    algorithm algo1{uda1};
+
+    // 3 - Instantiate populations
+    population pop{prob, 24};
+    population pop1{prob, 24};
+    population pop2{prob, 24};
+
+    // 4 - Evolve the population
+    pop1 = algo1.evolve(pop);
+
+    // 5 new algorithm that is bfe enabled
+    nsga2 uda2{nsga2{100}};
+    uda2.set_verbosity(1u);
+    uda2.set_seed(23u);
+    uda2.set_bfe(bfe{}); // This will use the default bfe.
+    // 6 - Instantiate a pagmo algorithm
+    algorithm algo2{uda2};
+
+    // 7 - Evolve the population
+    pop2 = algo2.evolve(pop);
+    BOOST_CHECK(uda1.get_log() == uda2.get_log());
```

Comment:
I am not sure here you are comparing what you think you are comparing. I think the udas need to be extracted from the algorithm, these are copies with an empty log. (check and see for yourself, these logs should both be empty thats why they are equal?)

Reply:
Sorry for not responding sooner,  thank you for taking the time to check this carefully.

I've tried to google boost testing tutorials, but I can't work out how to apply this in pagmo.  Where do the tests in pagmo run from?  As it stands I tried running it in a separate file (https://gist.github.com/willcodeforfudd/b66e5097a072ff6989acb7bd1d0c2ba6)  Comparing populations manually, I saw that they were the same,  so I assumed that the test was working due to confirmation bias (as you say, it isn't the test I thought it was) 

Do you propose to compare the algorithm logs, or would comparing population decision vectors and fitnesses work?

Reply:
@willcodeforfudd The issue here is that you are comparing the logs of the *original* algorithms, but these are not the algorithms used for evolutions (so their logs are empty). When you construct ``algo1`` and ``algo2``, copies of the original UDAs will be made and then used for performing the evolution. What @darioizzo is saying is that you should ask for the log from the UDAs stored within ``algo1`` and ``algo2``, something like:

```
BOOST_CHECK(algo1.extract<nsga2>()->get_log() == algo2.extract<nsga2>()->get_log());
```

With that said, comparing the populations would probably be good as well.

Reply:
OK,  I'm happy to fix this to the way you guys suggest.  Could you let me know how to run the tests?  I shouldn't really have submitted something without running it -  I'm afraid that my enthusiasm got the better of me.

Reply:
@willcodeforfudd in order to run the tests, turn on the options ``PAGMO_BUILD_PAGMO`` and ``PAGMO_BUILD_TESTS`` in CMake. Also make sure to set the build type to ``Debug``. Then, run the CMake build as usual (e.g., ``cmake --build .`` or something similar), the test executables will be in the ``tests/`` subdirectory of your build directory.

Reply:
Great,  thanks for explaining that,  I haddn't thought to go through ccmake for clues.  I'll run the tests and check the output.  I have now sorted out the bug I found earlier in nsga2, sorry for the inconvenience.  Thanks for your help.

Reply:
Alright,  I will have a look at the CI and try building the tests locally.  It looks like I put in the PR prematurely,  I guess there is a lesson here - of course I couldn't have found (some of) these issues without the CI and your help, so many thanks gents.

Reply:
This page says that the CI failed, but when I click through to Travis CI, it seems to think that it succeeded.  Running the test for nsga2 locally it claims to work ok.  How can I fix this?

Reply:
@willcodeforfudd the CI failure looked like a transient error, I restarted that specific build and it passed.

Reply:
Great,  thanks!  As far as I understand it, the change request is satisfied then?

Reply:
Yes merging this.